test/e2e/common/expansion.go

```diff
@@ -306,6 +307,137 @@ var _ = framework.KubeDescribe("Variable Expansion", func() {
+
+		podName := "var-expansion-" + string(uuid.NewUUID())
+		containerName := "dapi-container"
+		pod := &v1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:        podName,
+				Labels:      map[string]string{"name": podName},
+				Annotations: map[string]string{"notmysubpath": "mypath"},
+			},
+			Spec: v1.PodSpec{
+				Containers: []v1.Container{
+					{
+						Name:    containerName,
+						Image:   imageutils.GetE2EImage(imageutils.BusyBox),
+						Command: []string{"sh", "-c", "tail -f /dev/null"},
+						Env: []v1.EnvVar{
+							{
+								Name:  "POD_NAME",
+								Value: "foo",
+							},
+							{
+								Name: "ANNOTATION",
+								ValueFrom: &v1.EnvVarSource{
+									FieldRef: &v1.ObjectFieldSelector{
+										APIVersion: "v1",
+										FieldPath:  "metadata.annotations['mysubpath']",
+									},
+								},
+							},
+						},
+						VolumeMounts: []v1.VolumeMount{
+							{
+								Name:        "workdir1",
+								MountPath:   "/logscontainer",
+								SubPathExpr: "$(ANNOTATION)/$(POD_NAME)",
+							},
+							{
+								Name:      "workdir2",
+								MountPath: "/testcontainer",
+							},
+						},
+					},
+				},
+				RestartPolicy: v1.RestartPolicyNever,
+				Volumes: []v1.Volume{
+					{
+						Name: "workdir1",
+						VolumeSource: v1.VolumeSource{
+							HostPath: &v1.HostPathVolumeSource{Path: "/tmp"},
+						},
+					},
+					{
+						Name: "workdir2",
+						VolumeSource: v1.VolumeSource{
+							HostPath: &v1.HostPathVolumeSource{Path: "/tmp"},
+						},
+					},
+				},
+			},
+		}
+
+		By("creating the pod with failed condition")
+		pod, err := f.ClientSet.CoreV1().Pods(f.Namespace.Name).Create(pod)
+		Expect(err).ToNot(HaveOccurred(), "while creating pod")
+
+		err = framework.WaitTimeoutForPodRunningInNamespace(f.ClientSet, pod.Name, pod.Namespace, framework.PodStartShortTimeout)
+		Expect(err).To(HaveOccurred(), "while waiting for pod to be running")
+
+		var podClient *framework.PodClient
+		podClient = f.PodClient()
+
+		By("updating the pod")
+		podClient.Update(podName, func(pod *v1.Pod) {
+			pod.ObjectMeta.Annotations = map[string]string{"mysubpath": "mypath"}
+		})
+
+		By("waiting for pod running")
+		err = framework.WaitTimeoutForPodRunningInNamespace(f.ClientSet, pod.Name, pod.Namespace, framework.PodStartShortTimeout)
+		Expect(err).NotTo(HaveOccurred(), "while waiting for pod to be running")
+
+		By("creating a file in subpath")
+		cmd := "touch /testcontainer/mypath/foo/test.log"
+		_, err = framework.RunHostCmd(pod.Namespace, pod.Name, cmd)
+		if err != nil {
+			framework.Failf("expected to be able to write to subpath")
+		}
+
+		By("test for file in mounted path")
+		cmd = "test -f /logscontainer/test.log"
+		_, err = framework.RunHostCmd(pod.Namespace, pod.Name, cmd)
+		if err != nil {
+			framework.Failf("expected to be able to verify file")
+		}
+
+		By("updating the annotation value")
+		podClient.Update(podName, func(pod *v1.Pod) {
```

Comment:
I don't think this will actually restart the container.  I don't know of a simple way to trigger a container restart from a test.  For other subpath tests, we used liveness probes to trigger a container restart: https://sourcegraph.com/github.com/kubernetes/kubernetes/-/blob/test/e2e/storage/testsuites/subpath.go#L685

Reply:
That looks like a very intricate rewrite of the suite but this is what I observe
```
centos7:[root@localhost kubernetes]$ kubectl get pods -n var-expansion-3206 --watch
NAME                                                 READY     STATUS                       RESTARTS   AGE
var-expansion-170a303a-18c2-11e9-8f7a-525400225b53   0/1       CreateContainerConfigError   0          27s
var-expansion-170a303a-18c2-11e9-8f7a-525400225b53   0/1       CreateContainerConfigError   0         2m
var-expansion-170a303a-18c2-11e9-8f7a-525400225b53   1/1       Running   0         2m
var-expansion-170a303a-18c2-11e9-8f7a-525400225b53   1/1       Running   0         2m
var-expansion-170a303a-18c2-11e9-8f7a-525400225b53   1/1       Terminating   0         2m
^Ccentos7:[root@localhost kubernetes]$ kubectl get pods -n var-expansion-3206 -o yaml var-expansion-170a303a-18c2-11e9-8f7a-5254002b53
apiVersion: v1
kind: Pod
metadata:
  annotations:
    mysubpath: mynewpath
```
The final pod state before the namespace gets deleted, has the new path in it and doesn't the fact that the pod becomes ready after it was in an initial failing state indicates that it got restarted?

Reply:
So there's two scenarios we should test (and maybe actually have two separate test cases instead of combining it into one):

1) Pod is started with a missing env var.  Pod startup should fail until the env variable is corrected.
2) Pod is started successfully.  The env variable value changes.  A container restart should *not* remount the subpath with the new value. Note that container restart is different from pod restart (deletion and creation).

Reply:
Sorry I dont understand. I thought that was exactly what I was doing?
Do you mean just split this into 2 e2e tests? Or is it not doing what it is supposed to.

Reply:
I'm suggesting we split it out into two tests for clarity.  And also, I don't think updating the Pod the second time will trigger a container restart (I manually tried this).  I think the output you are showing is the first scenario from pod startup error -> success.

Reply:
(Just because the pod annotation got updated doesn't mean that the actual volume mount was updated)

Reply:
So how do you suggest we do the container restart? I can try and duplicate the subpath functions if thats going to do it?

Reply:
I'm fine with making them public

Reply:
It seems like that container  restart does everything needed in one hit, so I have now split the e2e tests into 3 separate components and this I think should satisfy everything now